magma/syntax/inline_combinational.py

```diff
@@ -84,9 +102,15 @@ def exec(self, etree, stree, env):
         result = fn()
         if not isinstance(result, tuple):
             result = (result,)
-        for key, value in zip(self.target_map.keys(), result):
+        for key, value in zip(self.wire_map.keys(), result):
             # Eval original target in env to do wiring, not sure if there's a
             # way we can avoid having to do eval here (other option is to
             # codegen the original augassign in a new tree, but that's
             # effectively the same as evaling)
             wire(eval(astor.to_source(mutable(key)).rstrip(), {}, env), value)
+        for key, value in zip(self.assign_map.keys(),
+                              result[len(self.wire_map):]):
+            value_name = gen_free_name(ast.NameConstant(False), env)
+            env.locals.update({value_name: value})
+            exec(f"{astor.to_source(mutable(key)).rstrip()} = {value_name}",
```

Comment:
The more evaling / execing there is the more I am afraid of our attempts at recreating the env will break.  Maybe this could move to a pattern similar to peak where users must pass locals / globals. E.g. https://github.com/cdonovick/peak/blob/master/examples/riscv/sim.py#L23

Having to write `m.comb(locals(), globals())` would be somewhat annoying but it rules out a very hard to debug class of bugs.

Reply:
Yea I think that's a reasonable idea, let me see if there's any safer way to do this.  I looked into rewriting from the enclosing scope but it's not so easy.  The main problem is that the raw source code is actually invalid (for the assign case, since you'll get a name error when referring to a variable assigned inside the inline_comb. block), so we can't use a decorator pattern on the class even if we wanted to (not ideal anyways since we'd have to decorate instead of subclass circuit).  Other options face similar issues because of the potential for a name error from the invalid code.

I'm going to think about whether there's an alternative syntax that avoids the name error, perhaps using a `with m.inline_combinational as syntax:` block could work.

Reply:
Basically if we can find a way to let the code evaluate without an exception, we could then just hijack part of the class pipeline to introspect the source, do the rewrite, then return the evaluation of the rewrite (but this has other potential problems, as we know with getting source from a class).